Refactoring font-face src parsing

vitorroriz opened the pull request:
#### ef82a019637be51b1882305e819eeb2bba96f316
<pre>
Refactoring font-face src parsing
<a href="https://bugs.webkit.org/show_bug.cgi?id=250471">https://bugs.webkit.org/show_bug.cgi?id=250471</a>
rdar://104135235

Reviewed by Myles C. Maxfield.

Refactoring font-face src parsing such that we parse each component
separated by commas, individually.

If all components can&apos;t be parsed, then the src descriptor is
invalidated. This was already done before, but now each component parser,
such as consumeFontFaceSrcURI() and consumeFontFaceSrcLocal() will
receive a subrange, and the whole subrange has to be parsed for the
component to be valid.  This will fix test failures we had when parsing
format() in a URI component.

The optional format() token can optionally follow a url() token. Before,
if garbage was following the url() token, we would ignore it and still
validate the component with an empty format. However, we would not
consume the range being parsed, and if the next token was not a Comma,
we would invalidate the whole descriptor.

* LayoutTests/fast/css/font-face-src-parsing-expected.txt:
Update on which src descriptors are valid.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-face-src-format-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-face-src-list-expected.txt:
Missing failures should pass now.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-face-src-tech-expected.txt:
We don&apos;t support tech() yet but we need to rebaseline here. Increment
tests will pass now.

* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontFaceSrcURI):
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontFaceSrcLocal):
Component is invalidated if the whole range is not consumed.
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontFaceSrc):

* LayoutTests/fast/css/font-face-src-parsing-expected.txt:
* LayoutTests/fast/css/font-face-src-parsing.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-face-src-format-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-face-src-list-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-face-src-tech-expected.txt:
* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontFaceSrcURI):
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontFaceSrcLocal):
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontFaceSrc):

Canonical link: <a href="https://commits.webkit.org/258870@main">https://commits.webkit.org/258870@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1d59f711a28ed6df6a049e4d79e5eba219f7a107

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103211 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12336 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/36208 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112458 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172655 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107164 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13361 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3240 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95436 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/110659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108985 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/10258 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/36208 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37892 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92080 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/36208 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79622 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5739 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/36208 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5910 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2851 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11899 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/36208 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6096 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7658 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
```diff
@@ -223,6 +223,8 @@ static RefPtr<CSSFontFaceSrcResourceValue> consumeFontFaceSrcURI(CSSParserTokenR
             return nullptr;
         format = arg.value().toString();
     }
+    if (!range.atEnd())
+        return nullptr;
```
Is this tested?

vitorroriz replied:
yes, I believe we should invalidate a component (but not the whole descriptor) if, at the end of parsing, the range is not at the end, meaning we haven't consumed all of it.  For example:
` { src: 'url("foo.ttf") dummy(xyzzy), url("bar.html")', valid: true },`
For this descriptor, the first component should be invalidated. For that, after consuming `url()`, we will just consume the next non-blank token if it's a `format()` (and in the future `tech()`). At the end of it, if the whole component was not consumed, this means that some garbage data was present and the component should be invalid.

Oh, while I wrote that I realized you are probably asking if we actually have a test for it, instead of if this behavior is intended. The fragment above comes from wpt font-face-src-format.html, which should be valid because of the second component and not the first. So we probably should add a subtest with the first component only, and valid:false. I'll create a separated PR and export it to WPT. We do have an internal fast test for it though at fast/css/font-face-src-parsing.html:

```
@font-face {
    src: url(foo.ttf) "invalid";
}
```

litherum commented on Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
```diff
@@ -232,13 +234,13 @@ static RefPtr<CSSValue> consumeFontFaceSrcLocal(CSSParserTokenRange& range)
     if (args.peek().type() == StringToken) {
         auto& arg = args.consumeIncludingWhitespace();
-        if (!args.atEnd())
+        if (!args.atEnd() || !range.atEnd())
```
Ditto

vitorroriz replied:
I'll create new subtests in the wpt tests in a separated PR:
https://bugs.webkit.org/show_bug.cgi?id=250566

litherum commented on Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
```diff
@@ -232,13 +234,13 @@ static RefPtr<CSSValue> consumeFontFaceSrcLocal(CSSParserTokenRange& range)
     if (args.peek().type() == IdentToken) {
         AtomString familyName = CSSPropertyParserHelpers::concatenateFamilyName(args);
-        if (familyName.isNull() || !args.atEnd())
+        if (familyName.isNull() || !args.atEnd() || !range.atEnd())
```
Ditto

vitorroriz replied:
I'll create new subtests in the wpt tests in a separated PR:
https://bugs.webkit.org/show_bug.cgi?id=250566

litherum commented on Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
```diff
@@ -249,16 +251,27 @@ RefPtr<CSSValueList> consumeFontFaceSrc(CSSParserTokenRange& range, const CSSPar
+            range.consumeComponentValue();
+
+        auto subrange = range.makeSubRange(begin, &range.peek());
+        auto parsedValue = consumeSrcListComponent(subrange);
```
```
if (auto parsedValue = consumeSrcListComponent(subrange))
```

litherum commented on Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
```diff
@@ -249,16 +251,27 @@ RefPtr<CSSValueList> consumeFontFaceSrc(CSSParserTokenRange& range, const CSSPar
-    } while (CSSPropertyParserHelpers::consumeCommaIncludingWhitespace(range));
+
+        if (!range.atEnd())
+            range.consumeIncludingWhitespace();
```
Why not `consumeComponentValue()`, as above?

vitorroriz replied:
The first `consumeComponentValue()` is for generating a subrange according to the comma delimiters. This one, is to remove any empty space in the beginning of a new component/subrange.

vitorroriz committed:
Refactoring font-face src parsing
https://bugs.webkit.org/show_bug.cgi?id=250471
rdar://104135235

Reviewed by Myles C. Maxfield.

Refactoring font-face src parsing such that we parse each component
separated by commas, individually.

If all components can't be parsed, then the src descriptor is
invalidated. This was already done before, but now each component parser,
such as consumeFontFaceSrcURI() and consumeFontFaceSrcLocal() will
receive a subrange, and the whole subrange has to be parsed for the
component to be valid.  This will fix test failures we had when parsing
format() in a URI component.

The optional format() token can optionally follow a url() token. Before,
if garbage was following the url() token, we would ignore it and still
validate the component with an empty format. However, we would not
consume the range being parsed, and if the next token was not a Comma,
we would invalidate the whole descriptor.

* LayoutTests/fast/css/font-face-src-parsing-expected.txt:
Update on which src descriptors are valid.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-face-src-format-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-face-src-list-expected.txt:
Missing failures should pass now.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-face-src-tech-expected.txt:
We don't support tech() yet but we need to rebaseline here. Increment
tests will pass now.

* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontFaceSrcURI):
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontFaceSrcLocal):
Component is invalidated if the whole range is not consumed.
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontFaceSrc):

* LayoutTests/fast/css/font-face-src-parsing-expected.txt:
* LayoutTests/fast/css/font-face-src-parsing.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-face-src-format-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-face-src-list-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-face-src-tech-expected.txt:
* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontFaceSrcURI):
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontFaceSrcLocal):
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontFaceSrc):

Canonical link: https://commits.webkit.org/258870@main
